## Import dependencies

In [99]:
"""
WEB SCRAPING TORONTO DATA FROM WIKIPEDIA
GETTING GEOLOCATION DATA 
ANALYSING NEIGHBOURHOOD DATA
"""
import pandas as pd
import requests
import bs4 as bs
import geocoder
import numpy as np
import folium
import math

import json
from pandas.io.json import json_normalize
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

# WEB SCRAPING TORONTO DATA FROM WIKIPEDIA

Define the url of the wikipedia page and load the relevant table into a beautiful soup object

In [25]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

#download a webpage as html data in a scrape with:
Web = requests.get(url)

#html content can then be analysed with:
soup = bs.BeautifulSoup(Web.text, "html.parser")


table_contents=[]
table=soup.find('table')

Parse the table to collect relevant information for our dataframe

In [26]:
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

#print(table_contents)

Convert the table into a dataframe 

In [27]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

# GETTING GEOLOCATION DATA 

Collecting the data using the geocoder package. The package does not fetch any data. The code to use it has been included but instead the csv is used.

In [28]:
"""
Geocoder version, wouldnt complete API call so alternative version is below.
"""
# #Get the list of postal code information
# postal_list = list(df["PostalCode"])

# latitude = []
# longitude = []
# #DO THIS IN A LOOP FOR EVERY POSTAL CODE
# for i in range(len(postal_list)):
#     postal_code = postal_list[i]
#     # initialize your variable to None
#     lat_lng_coords = None
    
#     # loop until you get the coordinates
#     while(lat_lng_coords is None):
#       g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#       lat_lng_coords = g.latlng
   
#     #Assign the coordinates to the dataframe
#     latitude.append(lat_lng_coords[0])
#     longitude.append(lat_lng_coords[1])
    
# df["lat"] = latitude
# df["long"] = longitude

# #Get the list of postal code information
# postal_list = list(df["PostalCode"])
"""
Reading the latlong from the csv provided and adding to our dataframe
"""
geo_df = pd.read_csv("/Users/stanleyrichards/Documents/Code/Notebooks/Geospatial_Coordinates.csv")

geo_df

postal_list = list(df["PostalCode"])
latitude = []
longitude = []
#DO THIS IN A LOOP FOR EVERY POSTAL CODE
for i in range(len(postal_list)):
    postal_code = postal_list[i]
    # Read the csv for the latlong
    lat = geo_df.loc[geo_df['Postal Code'] == postal_code, 'Latitude'].values[0]
    long = geo_df.loc[geo_df['Postal Code'] == postal_code, 'Longitude'].values[0]
    lat_lng_coords = [lat, long]
    
    
    #Assign the coordinates to the dataframe
    latitude.append(lat_lng_coords[0])
    longitude.append(lat_lng_coords[1])
    
df["Latitude"] = latitude
df["Longitude"] = longitude
pd.set_option('display.max_rows', None)
df

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

# EXPLORING AND CLUSTERING NEIGHBORHOODS 

# Exploring Neighborhoods in Toronto

In [29]:
address = 'Toronto, Ontario'

location = geocoder.google(address)

latitude = location.latitude
longitude = location.longitude
if latitude == None:
    #Then geocoder is again not working and we will just have to do it manually
    longitude = 79.3832
    latitude = 43.6532
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6532, 79.3832.


In [30]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df['Latitude'], df['Longitude'], df['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [31]:
CLIENT_ID = '54IF2K3CPOYH1NM4ZUWRMZVYI53TJFRKYB3VASQOXN5PMU3P' # your Foursquare ID
CLIENT_SECRET = 'YAEJEAZYCIB5IY413540YA3DC3F2JM5IPGEBCOCP5GAOBWIF' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 54IF2K3CPOYH1NM4ZUWRMZVYI53TJFRKYB3VASQOXN5PMU3P
CLIENT_SECRET:YAEJEAZYCIB5IY413540YA3DC3F2JM5IPGEBCOCP5GAOBWIF


# Analyse the first neighborhood

Getting the name and location details of the first neighborhood from the dataframe created above

In [32]:
print("The first neighborhood is " + df.loc[0, 'Neighborhood'])

neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

The first neighborhood is Parkwoods
Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


Creating a foursquare api access URL

In [33]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius


url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=54IF2K3CPOYH1NM4ZUWRMZVYI53TJFRKYB3VASQOXN5PMU3P&client_secret=YAEJEAZYCIB5IY413540YA3DC3F2JM5IPGEBCOCP5GAOBWIF&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

Getting the results of the API call in json format and defining a function to parse this json into a dataframe.

In [34]:
results = requests.get(url).json()

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Creating a dataframe of events returned by the API call for this first neighborhood.

In [37]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-37-561c05f0fdd1>:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name            categories        lat        lng
0  Brookbanks Park                  Park  43.751976 -79.332140
1              KFC  Fast Food Restaurant  43.754387 -79.333021
2    Variety Store     Food & Drink Shop  43.751974 -79.333114

In [38]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

3 venues were returned by Foursquare.


This first neighborhood has few venues around it.

# Analyse All Neighborhoods

First defining a function to do the above analysis of nearby neighborhoods automatically.

In [39]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now using this function we can loop through the neighborhoods in our dataframe containing all venues returned.

In [40]:
toronto_venues = getNearbyVenues(df['Neighborhood'], df['Latitude'], df['Longitude'], radius=500)
toronto_venues.head()

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1   Fast Food Restaurant  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

In [45]:
print("A total of {} venues have been returned".format(toronto_venues.shape[0]))
print("Grouped by neighborhood:")
toronto_venues.groupby('Neighborhood').count()

A total of 2130 venues have been returned
Grouped by neighborhood:


Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  9   
Bathurst Manor, Wilson Heights, Downsview North                        23   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      25   
Berczy Park                                                            60   
Birch Cliff, Cliffside West                                             5   
Brockton, Parkdale Village, Exhibition Place                           23   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Caledonia-Fairbanks                                                     4   
Cedarbrae                                                               8   
Central Bay Street                                                     60   
Christie                                                               15   
Church and Wellesley                                                   77   
Clairville, Humberwood, Woodbine Downs, West Hu...                      5   
Clarks Corners, Tam O'Shanter, Sullivan                                13   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills North                                                         4   
Don Mills South                                                        19   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview Central                                                       3   
Downsview East                                                          3   
Downsview Northwest                                                     4   
Downsview West                                                          4   
Dufferin, Dovercourt Village                                           16   
Enclave of L4W                                                         13   
Enclave of M4L                                                         15   
Enclave of M5E                                                        100   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           66   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                         9   
Guildwood, Morningside, West Hill                                       9   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
Hillcrest Village                                                       5   
Humber Summit                                                           3   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         21   
Kennedy Park, Ionview, East Birchmount Park                             5   
Kensington Market, Chinatown, Grange Park                              60   
Kingsview Village, St. Phillips, Martin Grove G...                      3   
Lawrence Manor, Lawrence Heights                

In [47]:
print('There are {} unique venue categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 272 unique venue categories.


# Cluster Neighborhoods

To cluster the neighborhoods in Toronto we will compare the venue categories in each neighborhood.

In [72]:
#Turning the categorical variable into a one hot encoded numerical values
toronto_onehot = None
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.drop(columns = ['Neighborhood'], inplace=True)


# add neighborhood column back to dataframe in the first index
data_ins = toronto_venues['Neighborhood']
toronto_onehot.insert(0, 'Neighborhood', data_ins)

#Display resulting dataframe
toronto_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0         Parkwoods                  0               0                  0   
1         Parkwoods                  0               0                  0   
2         Parkwoods                  0               0                  0   
3  Victoria Village                  0               0                  0   
4  Victoria Village                  0               0                  0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0        0                   0               0                0   
1        0                   0               0                0   
2        0                   0               0                0   
3        0                   0               0                0   
4        0                   0               0                0   

   Airport Terminal  American Restaurant  ...  Truck Stop  \
0                 0                    0  ...           0   
1                 0                    0  ...           0   
2                 0                    0  ...           0   
3                 0                    0  ...           0   
4                 0                    0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                0         0          0            0              0   
1                0         0          0            0              0   
2                0         0          0            0              0   
3                0         0          0            0              0   
4                0         0          0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 272 columns]

This dataframe can then be grouped by neighborhood, taking the mean of occurence in each category. This results in 101 neighborhoods each with 272 mean occurences for the various venue categories.

In [76]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped
#toronto_grouped.shape

Neighborhood  Accessories Store  \
0                                            Agincourt           0.000000   
1                               Alderwood, Long Branch           0.000000   
2      Bathurst Manor, Wilson Heights, Downsview North           0.000000   
3                                      Bayview Village           0.000000   
4                    Bedford Park, Lawrence Manor East           0.000000   
5                                          Berczy Park           0.000000   
6                          Birch Cliff, Cliffside West           0.000000   
7         Brockton, Parkdale Village, Exhibition Place           0.000000   
8    CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
9                                  Caledonia-Fairbanks           0.000000   
10                                           Cedarbrae           0.000000   
11                                  Central Bay Street           0.000000   
12                                            Christie           0.000000   
13                                Church and Wellesley           0.000000   
14   Clairville, Humberwood, Woodbine Downs, West H...           0.000000   
15             Clarks Corners, Tam O'Shanter, Sullivan           0.000000   
16     Cliffside, Cliffcrest, Scarborough Village West           0.000000   
17                      Commerce Court, Victoria Hotel           0.000000   
18                                          Davisville           0.000000   
19                                    Davisville North           0.000000   
20    Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
21                                     Don Mills North           0.000000   
22                                     Don Mills South           0.000000   
23   Dorset Park, Wexford Heights, Scarborough Town...           0.000000   
24                                   Downsview Central           0.000000   
25                                      Downsview East           0.000000   
26                                 Downsview Northwest           0.000000   
27                                      Downsview West           0.000000   
28                        Dufferin, Dovercourt Village           0.000000   
29                                      Enclave of L4W           0.000000   
30                                      Enclave of M4L           0.000000   
31                                      Enclave of M5E           0.000000   
32   Eringate, Bloordale Gardens, Old Burnhamthorpe...           0.000000   
33                        Fairview, Henry Farm, Oriole           0.000000   
34              First Canadian Place, Underground city           0.000000   
35                            Forest Hill North & West           0.000000   
36                            Garden District, Ryerson           0.000000   
37                                           Glencairn           0.000000   
38                     Golden Mile, Clairlea, Oakridge           0.000000   
39                   Guildwood, Morningside, West Hill           0.000000   
40   Harbourfront East, Union Station, Toronto Islands           0.000000   
41                       High Park, The Junction South           0.000000   
42                                   Hillcrest Village           0.000000   
43                                       Humber Summit           0.000000   
44                                    Humberlea, Emery           0.000000   
45                                  Humewood-Cedarvale           0.000000   
46                      India Bazaar, The Beaches West           0.000000   
47         Kennedy Park, Ionview, East Birchmount Park           0.000000   
48           Kensington Market, Chinatown, Grange Park           0.000000   
49   Kingsview Village, St. Phillips, Martin Grove ...           0.000000   
50                    Lawrence Manor, Lawrence Heights           0.090909   
51                                    

In [78]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0             Breakfast Spot  0.25
1                     Lounge  0.25
2  Latin American Restaurant  0.25
3             Clothing Store  0.25
4          Accessories Store  0.00


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place  0.22
1  Athletics & Sports  0.11
2         Coffee Shop  0.11
3                 Pub  0.11
4      Sandwich Place  0.11


----Bathurst Manor, Wilson Heights, Downsview North----
                venue  freq
0                Bank  0.09
1         Coffee Shop  0.09
2   Mobile Phone Shop  0.04
3  Chinese Restaurant  0.04
4         Supermarket  0.04


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1                 Bank  0.25
2   Chinese Restaurant  0.25
3                 Café  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0          Restaurant  0.08
1         Coffee Shop  0.08
2      Sandwich P

4  Furniture / Home Store  0.06


----Little Portugal, Trinity----
                           venue  freq
0                            Bar  0.09
1               Asian Restaurant  0.05
2          Vietnamese Restaurant  0.05
3                     Restaurant  0.05
4  Vegetarian / Vegan Restaurant  0.05


----Malvern, Rouge----
                  venue  freq
0  Fast Food Restaurant   0.5
1            Print Shop   0.5
2                Market   0.0
3   Martial Arts School   0.0
4        Massage Studio   0.0


----Milliken, Agincourt North, Steeles East, L'Amoreaux East----
                       venue  freq
0               Intersection  0.33
1                 Playground  0.33
2                       Park  0.33
3  Middle Eastern Restaurant  0.00
4        Moroccan Restaurant  0.00


----Mimico NW, The Queensway West, South of Bloor, Kingsway Park South West, Royal York South West----
                    venue  freq
0          Hardware Store  0.07
1  Thrift / Vintage Store  0.07
2          Disco

4    Monument / Landmark  0.00


----Woodbine Heights----
          venue  freq
0  Skating Rink  0.22
1    Beer Store  0.11
2      Bus Stop  0.11
3          Park  0.11
4  Intersection  0.11


----York Mills West----
                        venue  freq
0  Construction & Landscaping  0.33
1           Convenience Store  0.33
2                        Park  0.33
3           Accessories Store  0.00
4   Middle Eastern Restaurant  0.00


----York Mills, Silver Hills----
                 venue  freq
0  Martial Arts School   1.0
1    Accessories Store   0.0
2                Motel   0.0
3       Massage Studio   0.0
4       Medical Center   0.0




In [82]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt        Clothing Store   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot                Lounge   
1                        Gym              Pharmacy            Playground   
2                       Bank         Grocery Store   Fried Chicken Joint   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4         Italian Restaurant            Restaurant           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Event Space    Falafel Restaurant  Ethiopian Restaurant   
1           Coffee Shop    Athletics & Sports        Sandwich Place   
2           Bridal Shop        Sandwich Place                 Diner   
3        Discount Store   Distribution Center               Dog Run   
4      Greek Restaurant  Fast Food Restaurant     Indian Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Escape Room     Electronics Store         Discount Store  
1                   Pub   Distribution Center     Dim Sum Restaurant  
2            Restaurant          Intersection            Supermarket  
3      Doner Restaurant            Donut Shop            Yoga Studio  
4      Sushi Restaurant                   Pub           Liquor Store

# Clustering Neighborhoods

Using K-Means clustering to cluster the neighborhoods.

In [83]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 1], dtype=int32)

In [86]:
# # add clustering labels
# neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# toronto_merged = df

# # merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
# toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656             1.0     Food & Drink Shop    Fast Food Restaurant   
1 -79.315572             2.0           Pizza Place             Coffee Shop   
2 -79.360636             2.0           Coffee Shop                    Park   
3 -79.464763             2.0        Clothing Store  Furniture / Home Store   
4 -79.389494             2.0           Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
0                  Park    Dumpling Restaurant        Discount Store   
1          Hockey Arena  Portuguese Restaurant           Yoga Studio   
2                Bakery         Breakfast Spot                  Café   
3              Boutique  Vietnamese Restaurant    Miscellaneous Shop   
4                  Bank                    Bar              Beer Bar   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0   Distribution Center               Dog Run             Doner Restaurant   
1            Donut Shop                 Diner               Discount Store   
2                   Pub               Theater        Performing Arts Venue   
3           Coffee Shop     Accessories Store  Eastern European Restaurant   
4                   Spa         Smoothie Shop               Sandwich Place   

  9th Most Common Venue 10th Most Common Venue  
0            Donut Shop              Drugstore  
1   Distribution Center                Dog Run  
2            Shoe Store             Restaurant  
3     Electronics Store    Dumpling Restaurant  
4           Salad Place          Burrito Place

# Visualise the clusters produced by the K-Means algorithm

In [101]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    if math.isnan(cluster):
        idx = 0
    else:
        idx = int(cluster)-1
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[idx],
        fill=True,
        fill_color=rainbow[idx],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examining these clusters in more detail

In [103]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             0.0   Martial Arts School           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45   Dumpling Restaurant        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue       10th Most Common Venue  
45             Drugstore  Eastern European Restaurant

In [104]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             1.0     Food & Drink Shop   
21                    York             1.0                  Park   
35  East York/East Toronto             1.0                  Park   
40              North York             1.0                  Park   
52              North York             1.0                  Park   
64                    York             1.0                  Park   
66              North York             1.0                  Park   
77               Etobicoke             1.0                  Park   
85             Scarborough             1.0          Intersection   
91        Downtown Toronto             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Fast Food Restaurant                  Park   Dumpling Restaurant   
21               Women's Store                  Pool           Yoga Studio   
35           Convenience Store           Yoga Studio   Dumpling Restaurant   
40                     Airport      Business Service           Yoga Studio   
52                 Yoga Studio   Dumpling Restaurant        Discount Store   
64           Convenience Store           Yoga Studio   Dumpling Restaurant   
66  Construction & Landscaping     Convenience Store           Yoga Studio   
77           Mobile Phone Shop        Sandwich Place           Yoga Studio   
85                        Park            Playground             Drugstore   
91                  Playground                 Trail            Donut Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Discount Store   Distribution Center               Dog Run   
21             Drugstore                 Diner        Discount Store   
35        Discount Store   Distribution Center               Dog Run   
40   Dumpling Restaurant   Distribution Center               Dog Run   
52   Distribution Center               Dog Run      Doner Restaurant   
64        Discount Store   Distribution Center               Dog Run   
66   Dumpling Restaurant   Distribution Center               Dog Run   
77            Donut Shop        Discount Store   Distribution Center   
85                 Diner        Discount Store   Distribution Center   
91                 Diner        Discount Store   Distribution Center   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0       Doner Restaurant            Donut Shop                    Drugstore  
21   Distribution Center               Dog Run             Doner Restaurant  
35      Doner Restaurant            Donut Shop                    Drugstore  
40      Doner Restaurant            Donut Shop                    Drugstore  
52            Donut Shop             Drugstore  Eastern European Restaurant  
64      Doner Restaurant            Donut Shop                    Drugstore  
66      Doner Restaurant            Donut Shop                    Drugstore  
77               Dog Run      Doner Restaurant                    Drugstore  
85               Dog Run      Doner Restaurant                   Donut Shop  
91               Dog Run      Doner Restaurant                    Drugstore

In [105]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1                North York             2.0                Pizza Place   
2          Downtown Toronto             2.0                Coffee Shop   
3                North York             2.0             Clothing Store   
4              Queen's Park             2.0                Coffee Shop   
6               Scarborough             2.0       Fast Food Restaurant   
7                North York             2.0                       Café   
8                 East York             2.0                Pizza Place   
9          Downtown Toronto             2.0                Coffee Shop   
10               North York             2.0                       Park   
11                Etobicoke             2.0                     Bakery   
13               North York             2.0                        Gym   
14                East York             2.0               Skating Rink   
15         Downtown Toronto             2.0                Coffee Shop   
16                     York             2.0                      Field   
17                Etobicoke             2.0                   Pharmacy   
18              Scarborough             2.0                 Restaurant   
19             East Toronto             2.0                       Park   
20         Downtown Toronto             2.0                Coffee Shop   
22              Scarborough             2.0                Coffee Shop   
23                East York             2.0        Sporting Goods Shop   
24         Downtown Toronto             2.0                Coffee Shop   
25         Downtown Toronto             2.0              Grocery Store   
26              Scarborough             2.0           Hakka Restaurant   
27               North York             2.0   Mediterranean Restaurant   
28               North York             2.0                Coffee Shop   
29                East York             2.0          Indian Restaurant   
30         Downtown Toronto             2.0                Coffee Shop   
31             West Toronto             2.0                   Pharmacy   
32              Scarborough             2.0                 Playground   
33               North York             2.0             Clothing Store   
34               North York             2.0     Furniture / Home Store   
36         Downtown Toronto             2.0                Coffee Shop   
37             West Toronto             2.0                        Bar   
38              Scarborough             2.0           Department Store   
39               North York             2.0                       Café   
41             East Toronto             2.0           Greek Restaurant   
42         Downtown Toronto             2.0                Coffee Shop   
43             West Toronto             2.0                       Café   
44              Scarborough             2.0                   Bus Line   
46               North York             2.0                       Park   
47             East Toronto             2.0                       Park   
48         Downtown Toronto             2.0                Coffee Shop   
49               North York             2.0                       Park   
50               North York             2.0                        Gym   
51              Scarborough             2.0        American Restaurant   
53               North York             2.0                 Food Truck   
54             East Toronto             2.0                Coffee Shop   
55               North York             2.0             Sandwich Place   
56                     York             2.0             Discount Store   
58              Scarborough             2.0                       Farm   
59               North York             2.0                Coffee Shop   
60               North York             2.0       Gym / Fitness Center   
61          Central Toronto             2.0                       Park   
62          Central Toro

In [106]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57  North York             3.0        Baseball Field           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57             Drugstore        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
57   Dumpling Restaurant         Farmers Market

In [107]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12  Scarborough             4.0                   Bar           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
12             Drugstore        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
12               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
12   Dumpling Restaurant         Farmers Market